models/resnet/resnet.hpp

```diff
@@ -0,0 +1,485 @@
+   */
+  ResNet(std::tuple<size_t, size_t, size_t> inputShape,
+         const bool includeTop = true,
+         const bool preTrained = false,
+         const size_t numClasses = 1000);
+
+  //! Get Layers of the model.
+  ann::FFN<OutputLayerType, InitializationRuleType>&
+      GetModel() { return resNet; }
+
+  //! Load weights into the model and assumes the internal matrix to be
+  //  named "ResNet".
+  void LoadModel(const std::string& filePath);
+
+  //! Save weights for the model and assumes the internal matrix to be
+  //  named "ResNet".
+  void SaveModel(const std::string& filepath);
+
+ private:
+  /**
+   * Adds a Convolution Block depending on the configuration.
+   *
+   * @tparam SequentialType Layer type in which convolution block will
+   *     be added.
+   *
+   * @param inSize Number of input maps.
+   * @param outSize Number of output maps.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param kernelWidth Width of the filter/kernel.
+   * @param kernelHeight Height of the filter/kernel.
+   * @param padW Padding width of the input.
+   * @param padH Padding height of the input.
+   * @param downSample Bool if it's a downsample block or not. Default is false.
+   * @param downSampleInputWidth Input widht for downSample block.
+   * @param downSampleInputHeight Input height for downSample block.
+   */
+  template<typename SequentialType = ann::Sequential<>>
+  void ConvolutionBlock(SequentialType* baseLayer,
+                           const size_t inSize,
+                           const size_t outSize,
+                           const size_t strideWidth = 1,
+                           const size_t strideHeight = 1,
+                           const size_t kernelWidth = 3,
+                           const size_t kernelHeight = 3,
+                           const size_t padW = 1,
+                           const size_t padH = 1,
+                           const bool downSample = false,
+                           const size_t downSampleInputWidth = 0,
+                           const size_t downSampleInputHeight = 0)
+  {
+    size_t convInputWidth = inputWidth;
+    size_t convInputHeight = inputHeight;
+
+    if (downSample)
+    {
+      convInputWidth = downSampleInputWidth;
+      convInputHeight = downSampleInputHeight;
+    }
+
+    baseLayer->Add(new ann::Convolution<>(inSize, outSize, kernelWidth,
+        kernelHeight, strideWidth, strideHeight, padW, padH, convInputWidth,
+        convInputHeight));
+    mlpack::Log::Info << "Convolution: " << "(" << inSize << ", " <<
+        convInputWidth << ", " << convInputHeight << ")" << " ---> (";
+
+    if (!downSample)
+    {
+      // Updating input dimensions.
+      inputWidth = ConvOutSize(convInputWidth, kernelWidth, strideWidth, padW);
+      inputHeight = ConvOutSize(convInputHeight, kernelHeight, strideHeight,
+          padH);
+    }
```

Comment:
If we reset the input width and height above to downSampleWidth and downSampleHeight, then we can remove this if condition and always update inputWidth and inputHeight.

Reply:
I am not really sure what you meant by this. can you explain ?